File: 54. Spiral Matrix/54. Spiral Matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    const n = matrix.length
    const m = matrix[0].length
    let res = []
    let top =0 
    let left = 0
    let right = m -1
    let bottom = n-1 

    while (top<=bottom && left<=right){
        for (let i=left; i<=right;i++){
            res.push(matrix[top][i])
        }
        top++

        for (let i=top; i<=bottom;i++){
            res.push(matrix[i][right])
        }
        right--

        if (top<=bottom){
            for (let i=right; i>=left; i--){
                res.push(matrix[bottom][i])
            }
            bottom--
        }
        
        if (left<=right){
            for (let i=bottom; i>=top; i--){
                res.push(matrix[i][left])
            }
            left++
        }
        
    }
    return res
    
};

console.log(spiralOrder(matrix = [[1,2,3],[4,5,6],[7,8,9]]))

File: 54. Spiral Matrix/54. Spiral Matrix.py
from typing import List
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        
        n,m = len(matrix), len(matrix[0])
        res = []
        left,top = 0,0
        right,bottom = m-1 , n-1 

        while left<= right and top<=bottom:

            for i in range(left,right+1):
                res.append(matrix[top][i])
            top+=1

            for i in range(top,bottom+1):
                res.append(matrix[i][right])
            right-=1

            if top <= bottom:
                for i in reversed(range(left, right + 1)):
                    res.append(matrix[bottom][i])
                bottom -= 1
            
            if left<=right:
                for i in reversed(range(top, bottom+1)):
                    res.append(matrix[i][left])
                left+=1
        
        return res
